File: day05/aoc5-2.py
#!/usr/bin/env python3

INPUT_FILE = "aocdata.txt"

def read_ranges(path):
    ranges = []
    with open(path, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if line == "":
                break  # stop at blank line
            start, end = map(int, line.split('-'))
            ranges.append((start, end))
    return ranges

def merge_ranges(ranges):
    ranges.sort()
    merged = []
    for start, end in ranges:
        if not merged or start > merged[-1][1] + 1:
            merged.append([start, end])
        else:
            merged[-1][1] = max(merged[-1][1], end)
    return merged

def main():
    ranges = read_ranges(INPUT_FILE)
    merged = merge_ranges(ranges)
    total_fresh = sum(end - start + 1 for start, end in merged)
    print(f"Total fresh IDs: {total_fresh}")

if __name__ == "__main__":
    main()

File: day05/aoc5-2.pl
#!/usr/bin/env perl
use strict;
use warnings;

my $INPUT_FILE = "aocdata.txt";

open my $fh, '<', $INPUT_FILE or die "Cannot open $INPUT_FILE: $!";
my @ranges;
while (my $line = <$fh>) {
    chomp $line;
    last if $line =~ /^\s*$/;  # stop at blank line
    my ($start, $end) = split /-/, $line;
    push @ranges, [$start, $end];
}
close $fh;

# Sort ranges
@ranges = sort { $a->[0] <=> $b->[0] } @ranges;

# Merge
my @merged;
for my $r (@ranges) {
    my ($start, $end) = @$r;
    if (!@merged || $start > $merged[-1]->[1] + 1) {
        push @merged, [$start, $end];
    } else {
        $merged[-1]->[1] = $end if $end > $merged[-1]->[1];
    }
}

# Count
my $total = 0;
for my $m (@merged) {
    $total += $m->[1] - $m->[0] + 1;
}

print "Total fresh IDs: $total\n";
